### ISSUE TITLE
Thread pool exhaustion on version 11 of `Mindscape.Raygun4Net.AspNetCore`

### ISSUE THREAD

**Issue by olane** [MindscapeHQ/raygun4net#541]
**Describe the bug**
Since upgrading to version 11 we have been seeing our instances suffer from what we believe to be thread pool exhaustion. The problem is apparent on versions `11.0.2` and `11.0.3` of `Mindscape.Raygun4Net.AspNetCore` but does not appear when we downgraded back to our previous version (`9.0.4`).

This is manifesting in slower HTTP responses from one of our services, which we noticed when the service started failing liveness/ready probes. This is a fairly simple service that has a few endpoints and calls a few external APIs when hit, but generally doesn't do that much business logic.

On further inspection, we noticed that when instances are started they seem to be in a "good" state (no slowdowns, no liveness probe failures), but they can be flipped into a "bad" state (some slowdowns, persistent periodic liveness probe failures). The only other signifier of this bad state is that the thread count metrics were higher for these instances and so was the threadpool queue length. Furthermore, the timing of when instances got shifted into this bad state was correlated with when those instances sent their first raygun error.

We ran `dotnet-stack` on an affected instance, and the majority of the threads in progress (8 in total) had stacks like the following:
```
Thread (0x1DA1):
  [Native Frames]
  System.Private.CoreLib!System.Threading.SemaphoreSlim.WaitUntilCountOrTimeout(int32,unsigned int32,value class System.Threading.CancellationToken)
  System.Private.CoreLib!System.Threading.SemaphoreSlim.Wait(int32,value class System.Threading.CancellationToken)
  System.Collections.Concurrent!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)
  System.Collections.Concurrent!System.Collections.Concurrent.BlockingCollection`1+<GetConsumingEnumerable>d__68[System.__Canon].MoveNext()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Threading.Tasks.VoidTaskResult].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Threading.Tasks.VoidTaskResult].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpClient+<<SendAsync>g__Core|83_0>d.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpClient+<<SendAsync>g__Core|83_0>d].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.__Canon].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Http!System.Net.Http.RedirectHandler+<SendAsync>d__4.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.RedirectHandler+<SendAsync>d__4].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.DiagnosticsHandler+<SendAsyncCore>d__8.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.DiagnosticsHandler+<SendAsyncCore>d__8].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpConnectionPool+<SendWithVersionDetectionAndRetryAsync>d__84.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpConnectionPool+<SendWithVersionDetectionAndRetryAsync>d__84].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.__Canon].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Http!System.Net.Http.HttpConnection+<SendAsyncCore>d__64.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpConnection+<SendAsyncCore>d__64].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.Threading.Tasks.VoidTaskResult].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpConnection+<InitialFillAsync>d__101.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,System.Net.Http.HttpConnection+<InitialFillAsync>d__101].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Int32].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Security!System.Net.Security.SslStream+<ReadAsyncInternal>d__188`1[System.Net.Security.AsyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Int32,System.Net.Security.SslStream+<ReadAsyncInternal>d__188`1[System.Net.Security.AsyncReadWriteAdapter]].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Int32].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Security!System.Net.Security.SslStream+<EnsureFullTlsFrameAsync>d__186`1[System.Net.Security.AsyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Int32,System.Net.Security.SslStream+<EnsureFullTlsFrameAsync>d__186`1[System.Net.Security.AsyncReadWriteAdapter]].MoveNext(class System.Threading.Thread)
  System.Net.Sockets!System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.InvokeContinuation(class System.Action`1<class System.Object>,class System.Object,bool,bool)
  System.Net.Sockets!System.Net.Sockets.SocketAsyncEngine.System.Threading.IThreadPoolWorkItem.Execute()
  System.Private.CoreLib!System.Threading.ThreadPoolWorkQueue.Dispatch()
  System.Private.CoreLib!System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
```

Running `dotnet-trace` on the instance using the `cpu-sampler` profile indicated that there were 7 threads spending most of their time in Raygun code, presumably waiting on something.

At this point we downgraded the package and the problems have not reoccurred since.

We don't have any customisations to the base settings, we're initializing like this:

```
        services.AddRaygun(raygunSettings =>
        {
            raygunSettings.ApiKey = m_Configuration.GetValue<string>("RAYGUN_API_KEY");
        });
```

**Comment by phillip-haydon** [comment]
Hello, this is obviously very concerning and I would like dig into this more. 

I'm going to try reproduce your results, but I'm wondering if you're able to test v10.1.2

The major difference between 9 and 10 is that in asp.net we treat the RaygunClient as a singleton, so instead of creating the client every time an exception occurs, it will reuse the client. The changes in 9 and 10 were done to fix Port Exhaustion in Azure and Thread Exhaustion when 1000's of exceptions get raised in a very short period of time. 

The changes in 11 include PDB support and offline storage.

Since the queuing was introduced in 9, I would expect the stack shown above to be similar.

**Comment by phillip-haydon** [comment]
I've managed to reproduce your results and it does occur in 10.1.2, it is a side effect of treating the RaygunClient as a singleton, will work on a solution for this. Thanks.

**Comment by olane** [comment]
Thanks for the quick response on this - do let me know if there's anything more I can provide to help, but sounds like you've found the cause. I assume you don't need me to test `10.1.2` but I can do so if it would be useful.

**Comment by phillip-haydon** [comment]
Hello @olane, I've released v11.0.4-pre-2 if you wouldn't mind giving it a whirl and seeing if you still observe the issues you were having.

We are also pending some feedback from a couple of MAUI customers for pre-1 and this particular issue may resolve a GC issue we observed in MAUI so we need to test that too before doing a full release. 

Thanks.

**Comment by olane** [comment]
Thanks Phillip - will probably be some time next week before we can properly test but I will do ASAP.

**Comment by phillip-haydon** [comment]
Hello @olane I've done another pre-release `11.0.4-pre-4` to resolve a race condition. Plan is to do a full release on Monday unless you have any objections or issues.

**Comment by olane** [comment]
Thanks for the update. We have now had pre-2 running for about 24 hours and we are yet to see a recurrence of the issue we were seeing, so I would tentatively say it's fixed. Appreciate the quick turnaround on this one - I'll let you know if we see any other problems.

**Comment by phillip-haydon** [comment]
Thanks for the update! Happy to hear it’s working well so far!

**Comment by phillip-haydon** [comment]
This has been released as v11.1.0

Let us know if there's any further issues. Thanks!

**Comment by diepnt90** [comment]
I'm running 11.11.1 but still observed lots of threads with the same call stack as below
`<html>
<body>

GCFrameHelperMethodFrame_1OBJSystem.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutSystem.Threading.SemaphoreSlim.WaitSystem.Collections.Concurrent.BlockingCollection`1[[System.__Canon mscorlib]].TryTakeWithNoTimeValidationSystem.Collections.Concurrent.BlockingCollection`1+d__68[[System.__Canon mscorlib]].MoveNextMindscape.Raygun4Net.ThrottledBackgroundMessageProcessor.RaygunMessageWorkerSystem.Threading.Tasks.Task.ExecuteSystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.Tasks.Task.ExecuteWithThreadLocalSystem.Threading.Tasks.Task.ExecuteEntrySystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.ExecutionContext.RunSystem.Threading.ThreadHelper.ThreadStartGCFrameDebuggerU2MCatchHandlerFrameContextTransitionFrameDebuggerU2MCatchHandlerFrame | GCFrameHelperMethodFrame_1OBJSystem.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutSystem.Threading.SemaphoreSlim.WaitSystem.Collections.Concurrent.BlockingCollection`1[[System.__Canon mscorlib]].TryTakeWithNoTimeValidationSystem.Collections.Concurrent.BlockingCollection`1+d__68[[System.__Canon mscorlib]].MoveNextMindscape.Raygun4Net.ThrottledBackgroundMessageProcessor.RaygunMessageWorkerSystem.Threading.Tasks.Task.ExecuteSystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.Tasks.Task.ExecuteWithThreadLocalSystem.Threading.Tasks.Task.ExecuteEntrySystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.ExecutionContext.RunSystem.Threading.ThreadHelper.ThreadStartGCFrameDebuggerU2MCatchHandlerFrameContextTransitionFrameDebuggerU2MCatchHandlerFrame
-- | --
GCFrameHelperMethodFrame_1OBJSystem.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutSystem.Threading.SemaphoreSlim.WaitSystem.Collections.Concurrent.BlockingCollection`1[[System.__Canon mscorlib]].TryTakeWithNoTimeValidationSystem.Collections.Concurrent.BlockingCollection`1+d__68[[System.__Canon mscorlib]].MoveNextMindscape.Raygun4Net.ThrottledBackgroundMessageProcessor.RaygunMessageWorkerSystem.Threading.Tasks.Task.ExecuteSystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.Tasks.Task.ExecuteWithThreadLocalSystem.Threading.Tasks.Task.ExecuteEntrySystem.Threading.ExecutionContext.RunInternalSystem.Threading.ExecutionContext.RunSystem.Threading.ExecutionContext.RunSystem.Threading.ThreadHelper.ThreadStartGCFrameDebuggerU2MCatchHandlerFrameContextTransitionFrameDebuggerU2MCatchHandlerFrame


</body>
</html>`
